### PR TITLE
Hydrate all function dependencies in container

### PR DESCRIPTION
**Pull request by erikbern**
This lets us avoid stub assignments in many cases. Example:

```python
import modal

stub = modal.Stub()
volume = modal.Volume.new()

@stub.function(volumes={"/xyz": volume})
def run():
    with open("/xyz/abc.txt", "w") as f:
        f.write("hello")
    volume.commit()
```

This works because we track that the volume is a dependency of the function, and inside the container we hydrate the volume back. This works for _any_ object, including images (useful for `.run_inside`) etc.

Note that we currently don't track objects referred to inside of functions. It has to be an explicit dependency. So this won't work – yet:

```python
import modal

stub = modal.Stub()
queue = modal.Queue.new()

@stub.function()
def run():
    queue.push(123)
```

In order for this to work, we also need to add all `inspect.closurevars` of the function. I'll add that later. With these changes, we should be able to deprecate stub assignments.

### CODE REVIEW

**Review by aksh-at — APPROVED** [review]
Neat! Discussed implications of relying on ordering on Slack, but seems like a really unlikely edge case.

File: test/container_test.py [hunk 1]
```diff
@@ -546,7 +552,7 @@ def test_cli(unix_servicer, event_loop):
 
 
 @skip_windows_unix_socket
-def test_function_hydration(unix_servicer):
+def test_function_subling_hydration(unix_servicer):
```
[the remark this answers is not in the document]
Comment (erikbern) [reply test/container_test.py:555]: thanks will fix

**Comment by erikbern** [comment]
Thanks @aksh-at !
